planetiler-core/src/main/java/com/onthegomap/planetiler/util/LoopLineMerger.java

```diff
@@ -0,0 +1,438 @@
+          for (var loop : loops) {
+            if (loop != min) {
+              loop.edge.remove();
+            }
+          }
+        }
+      }
+    }
+  }
+
+  private double shortestDistance(Node start, Node end, Node exclude, double maxLength) {
+    Map<Integer, Double> bestDistance = new HashMap<>();
+    record Candidate(Node node, double cost, double heuristic) {}
+    PriorityQueue<Candidate> frontier = new PriorityQueue<>(Comparator.comparingDouble(Candidate::heuristic));
+    if (exclude != start) {
+      frontier.offer(new Candidate(start, 0, start.distance(end)));
+    }
+    while (!frontier.isEmpty()) {
+      Candidate candidate = frontier.poll();
+      Node current = candidate.node;
+      if (current == end) {
+        return candidate.cost;
+      }
+
+      for (var edge : current.getEdges()) {
+        var neighbor = edge.to;
+        if (neighbor != exclude) {
+          double newDist = candidate.cost + edge.length;
+          double prev = bestDistance.getOrDefault(neighbor.id, Double.POSITIVE_INFINITY);
+          if (newDist < prev) {
+            bestDistance.put(neighbor.id, newDist);
+            double heuristic = newDist + neighbor.distance(end);
+            if (heuristic <= maxLength) {
+              frontier.offer(new Candidate(neighbor, newDist, heuristic));
+            }
+          }
+        }
+      }
+    }
+    return Double.POSITIVE_INFINITY;
+  }
+
+  private void removeShortStubEdges(double stubMinLength) {
+    for (var node : output) {
+      for (var edge : List.copyOf(node.getEdges())) {
+        if (edge.length < stubMinLength &&
+          (edge.from.getEdges().size() == 1 || edge.to.getEdges().size() == 1 || edge.from == edge.to)) {
+          edge.remove();
+        }
+      }
+    }
+  }
+
+  private void removeShortEdges() {
+    for (var node : output) {
+      for (var edge : List.copyOf(node.getEdges())) {
+        if (edge.length < minLength) {
+          edge.remove();
+        }
+      }
+    }
+  }
+
+  public List<LineString> getMergedLineStrings() {
+    List<List<Coordinate>> edges = nodeLines(input);
+    buildNodes(edges);
+
+    merge();
+
+    if (loopMinLength > 0.0) {
+      breakLoops();
+      merge();
+    }
+
+    if (minLength > 0.0) {
+      double step = 1.0 / precisionModel.getScale();
+      for (double stubMinLength = 0.0; stubMinLength < minLength; stubMinLength += step) {
+        removeShortStubEdges(stubMinLength);
+        merge();
+      }
+      removeShortStubEdges(minLength);
+      merge();
+      // minLength = 10 * 0.0625;
+      // removeShortEdges();
+      // merge();
+    }
+
+    List<LineString> result = new ArrayList<>();
+
+    for (var node : output) {
+      for (var edge : node.getEdges()) {
+        if (edge.main) {
+          result.add(factory.createLineString(edge.coordinates.toArray(Coordinate[]::new)));
+        }
+      }
+    }
+
+    return result;
+  }
+
+  public List<LineString> getMergedByAngle() {
```

Comment:
let's remove this public API and make it a configurable part of regular merging.